ton_client/client/src/processing/process_message.rs

```diff
@@ -1,84 +1,98 @@
-use crate::abi::ParamsOfEncodeMessage;
-use crate::processing::types::{CallbackParams,  ProcessingOptions, TransactionResult};
+use crate::abi::{Abi, ParamsOfEncodeMessage};
+use crate::client::ClientContext;
+use crate::error::ApiResult;
+use crate::processing::internal::can_retry_expired_message;
+use crate::processing::types::{CallbackParams, TransactionOutput};
+use crate::processing::{
+    send_message, wait_for_transaction, ErrorCode, ParamsOfSendMessage, ParamsOfWaitForTransaction,
+    ResultOfWaitForTransaction,
+};
+use std::sync::Arc;
 
 #[derive(Serialize, Deserialize, TypeInfo, Debug)]
 pub enum MessageSource {
-    Message(String),
-    AbiEncoding(ParamsOfEncodeMessage),
+    Encoded { message: String, abi: Option<Abi> },
+    AbiEncodingParams(ParamsOfEncodeMessage),
 }
 
 #[derive(Serialize, Deserialize, TypeInfo, Debug)]
 pub struct ParamsOfProcessMessage {
     /// Message source.
     pub message: MessageSource,
-    /// Processing options.
-    pub processing_options: Option<ProcessingOptions>,
     /// Processing callback.
     pub callback: Option<CallbackParams>,
 }
 
-#[derive(Serialize, Deserialize, TypeInfo, PartialEq, Debug)]
-pub struct ResultOfProcessMessage {
-    pub transaction: Option<TransactionResult>,
-}
+/// Sends message to the network and monitors network for a result of
+/// message processing.
+#[method_info(name = "processing.process_message")]
+pub async fn process_message(
+    context: Arc<ClientContext>,
+    params: ParamsOfProcessMessage,
+) -> ApiResult<TransactionOutput> {
+    let abi = match &params.message {
+        MessageSource::Encoded { abi, .. } => abi.clone(),
+        MessageSource::AbiEncodingParams(encode_params) => Some(encode_params.abi.clone()),
+    };
+
+    let mut try_index = 0;
+    loop {
+        // Encode (or use encoded) message
+        let message = match &params.message {
+            MessageSource::Encoded { message, .. } => message.clone(),
```

Comment:
Is `Encoded` message retried as well as recreated from `AbiEncodingParams`? It shouldn't

Reply:
done